controllers/idler/aap_idler.go

```diff
@@ -0,0 +1,231 @@
+		podLogger := log.FromContext(ctx).WithValues("pod_name", pod.Name, "pod_phase", pod.Status.Phase)
+		podCtx := log.IntoContext(ctx, podLogger)
+
+		// check the restart count for the pod
+		restartCount := getHighestRestartCount(pod.Status)
+		if restartCount > AAPRestartThreshold {
+			podLogger.Info("Pod is restarting too often for an AAP pod. Checking if it belongs to AAP and if so then idle the aap", "restart_count", restartCount)
+			aapName, err = r.ensureAAPIdled(podCtx, pod, aapAPI)
+			if err != nil {
+				return err
+			}
+		} else {
+			// Check if running for longer then the AAP idler timeout
+			timeoutSeconds := r.aapTimeoutSeconds(idler)
+			if pod.Status.StartTime != nil && time.Now().After(pod.Status.StartTime.Add(time.Duration(timeoutSeconds)*time.Second)) {
+				podLogger.Info("Pod is running for too long for an AAP pod. Checking if it belongs to AAP and if so then idle the aap.", "start_time", pod.Status.StartTime.Format("2006-01-02T15:04:05Z"), "timeout_seconds", timeoutSeconds)
+				aapName, err = r.ensureAAPIdled(podCtx, pod, aapAPI)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		// Check if we need to send a notification and proceed with the next pod
+		if aapName != "" {
+			// The AAP has been Idled
+			// A notification should be sent
+			if err := r.createNotification(podCtx, idler, aapName, "Ansible Automation Platform"); err != nil {
+				podLogger.Error(err, "failed to create Notification")
+				if err = r.setStatusIdlerNotificationCreationFailed(podCtx, idler, err.Error()); err != nil {
+					podLogger.Error(err, "failed to set status IdlerNotificationCreationFailed")
+				} // not returning error to continue tracking remaining pods
+			}
+
+			idledAAPs[aapName] = aapName
+			if len(idledAAPs) == len(aapList.Items) {
+				// All AAPs are idled, no need to check the rest of the pods
+				return nil
+			}
+		}
+	}
+
+	return nil
+}
+
+const oneHour = 60 * 60 // in seconds
+
+func (r *Reconciler) aapTimeoutSeconds(idler *toolchainv1alpha1.Idler) int32 {
+	// Check if the idler timeout is less than one hour and if so, set it to half of the timeout.
+	// Otherwise, subtract one hour from the timeout.
+	// This is to ensure that the AAP idler kicks in before the main idler.
+	timeoutSeconds := idler.Spec.TimeoutSeconds
+	if timeoutSeconds <= oneHour {
+		timeoutSeconds = timeoutSeconds / 2
+	} else {
+		timeoutSeconds = timeoutSeconds - oneHour
+	}
+	return timeoutSeconds
+}
+
+// ensureAAPIdled checks if the long-running or crash-looping pod belongs to an AAP instance and if so, ensures that the AAP is idled.
+// Returns the AAP resource name in case it was idled, or an empty string if it was not idled.
+func (r *Reconciler) ensureAAPIdled(ctx context.Context, pod corev1.Pod, aapAPI aapAPI) (string, error) {
+	for _, podCond := range pod.Status.Conditions {
+		if podCond.Type == "PodCompleted" {
+			// Pod is in the completed state, no need to idle
+			return "", nil
+		}
+	}
```

Comment:
this code doesn't seem to be aap-specific, this should rather go into the main logic, correct?

Reply:
We actually want to delete the pod (which we do), but using the logic from the main idler.
I would simply drop it from here as it brings confusion why we have a special handling of these pods in the aap idler.